### PR TITLE
Adding UserStory4 Hotelzimmer Buchung

### LINKED ISSUE #14: U4 Hotelzimmer reservieren
> Als Gast möchte ich zu einem ausgewählten Zeitraum ein ausgewähltes Zimmer buchen, um eine Unterkunft in meinem Urlaub zu haben.

**Details**
Es besteht eine API Schnittstelle um eine Buchung zu einem Zimmer durchführen zu
können. Diese ist in der API Dokumentation zu finden.
Nachdem der Benutzer ein verfügbares Zimmer gefunden hat, soll er es Buchen können.
Er muss dazu folgende Daten angeben:
* Vorname
* Nachname
* gültige E-Mail-Adresse
* E-Mail-Adresse bestätigen
* Frühstück Ja / Nein
Er soll vor dem endgültigen Absenden, nochmal die Möglichkeit haben, seine
Buchungsdaten (inkl. Buchungszeitraum und Zimmerwahl) zu überprüfen und ggf. zu
bearbeiten.
Nach erfolgter Buchung muss dem Benutzer eine Bestätigung angezeigt werden.

**Recommended Approach**
* Machen Sie sich mit der Schnittstelle vertraut
* Analyse der zu übertragenden Daten
* Paperprototype inkl. Userflow und Errorhandling
* Bauen Sie zuerst den Store und die API Anbindung und dann das GUI

**Definition of Done**
* Der Benutzer kann zu einem gewünschten Zeitraum ein gewünschtes Zimmer unter Angabe seiner vollständigen und korrekten Daten Buchen, wenn das Zimmer verfügbar ist
* Mögliche Fehlerfälle werden abgefangen und der Benutzer entsprechend darauf hingewiesen
* Der Benutzer kann seine Daten vor dem Absenden nochmal überprüfen
* Der Benutzer erhält eine Bestätigung nach erfolgreicher Buchung
* Das Feature wurde in einem eigenen Branch eingecheckt und ins Git-Repository gepusht
* Das Feature wurde nach einem Code Review in den Master gemerged

### PR DESCRIPTION
**Pull request by StroSte**

### CODE REVIEW

**Comment by AntonijeStefanovic** [comment]
I booked room number 6 from July 7th to July 14th. The booking process worked correctly. However, I encountered an issue when checking availability for these dates afterward. Here are the steps to reproduce the issue:

1. Initial Availability Check:

- I checked the availability for room number 6 for the dates July 7th to July 14th.

- A popup correctly indicated that the room was not available.

2. Changing Dates:

- I changed the booking dates to some random dates in August.

- The form became available to fill in, as expected.

3. Returning to Unavailable Dates:

- I changed the dates back to the initially booked dates (July 7th to July 14th).

- The popup correctly indicated that the room was not available.

- However, the form remained active (not greyed out), allowing me to click on the "Buchen" button.

4. Error Encounter:

Clicking "Buchen" led to a 404 error in the console.

**Suggested Fix:
Ensure the form is disabled (greyed out) when the room is unavailable for the selected dates, and the "Buchen" button is not clickable.**

File: client/src/pages/RoomDetailsPage.vue [hunk 1]
```diff
@@ -1,25 +1,61 @@
...
+    ImageComponent,
+    CharacteristicsTable,
+    RoomAvailability,
   },
   data: () => ({
     roomStore: useRoomStore(),
+    bookingStore: useBookingStore(),
   }),
+  computed: {
+    bookingDisabled() {
```
Comment (AntonijeStefanovic) [inline client/src/pages/RoomDetailsPage.vue:49]: // ISSUE: The form should also be disabled when the room is not available, based on the availability check.
// The current implementation allows form submission even when the room is not available, leading to a 404 error.
https://github.com/Lammer-FH/boutique-hotel-technikum-2024-ae/pull/24#issuecomment-2176978688

File: client/src/store/useRoomStore.js [hunk 2]
```diff
@@ -59,8 +71,42 @@ export const useRoomStore = defineStore("room", {
...
+          .then((response) => {
+            if (
+              response.data &&
+              response.data.roomId === this.currentRoom.roomId
+            ) {
+              this.availableRoomId = response.data.roomId;
+            }
+          })
+          .catch((error) => {
+            console.log(error);
```
Comment (AntonijeStefanovic) [inline client/src/store/useRoomStore.js:103]: https://github.com/Lammer-FH/boutique-hotel-technikum-2024-ae/pull/24#discussion_r1645097008

File: server/boutique-hotel-technikum/src/main/java/com/example/boutiquehoteltechnikum/services/BookingService.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+			DateValidator.getDate(bookingRequestObject.getEndDate())
+		);
+		if (room == null)
+		{
+			throw new ResponseStatusException(
+				HttpStatus.NOT_FOUND, "Das Zimmer ist zu den angegebenen Daten nicht mehr verfügbar!"
+			);
+		}
+
+		GuestEntity guestEntity = guestRepository.save(GuestEntity.builder()
```
Comment (AntonijeStefanovic) [inline server/boutique-hotel-technikum/src/main/java/com/example/boutiquehoteltechnikum/services/BookingService.java:40]: // ISSUE: Users with the same email address cannot make multiple bookings?

Comment (StroSte) [reply server/boutique-hotel-technikum/src/main/java/com/example/boutiquehoteltechnikum/services/BookingService.java:40]: As the email is not a unique key, this should not be a problem

File: client/src/pages/RoomDetailsPage.vue (hunk 1 above)
Comment (StroSte) [reply client/src/pages/RoomDetailsPage.vue:49]: I tried to recreate the issues, but was not able to see it,

does any one else have the same kind of issue?

**Pull request merged**